posts/2017-03-01-four-ways-to-escape-a-cake.md

```diff
@@ -0,0 +1,476 @@
+```
+
+This has the proper signature, and it’s cleaner, since we don’t expose
+the unused-at-runtime `u` variable anymore. We could refine a little
+further, and replace it with a `U#Haystack`, just as with
+`twoInitsFromAHaystack`.
+
+This gives us the same interface, with all the index preservation we
+need. Even better, it infers a nicer return type.
+
+```scala
+scala> def trial = lu stepTwiceLively lu.haystack.init
+trial: lu.Needle
+```
+
+Now, let’s turn to implementation.
+
+```scala
+class OnceMoreStepTwice[U <: LittleUniverse](u: U) {
+  def stepTwiceFinally(n: U#Needle): U#Needle =
+    u.haystack.iter(u.haystack.iter(n))
+}
+
+<console>:18: error: type mismatch;
+ found   : U#Needle
+ required: OnceMoreStepTwice.this.u.Needle
+           u.haystack.iter(u.haystack.iter(n))
+                                           ^
+```
+
+This is the last part of the escape! If this worked, we could *fully
+erase* the `LittleUniverse` from most code, relying on the pure
+type-level index to prove enough of its existence! So it’s a little
+frustrating that it doesn’t quite work.
+
+Let’s break it down. First, the return type is fine.
+
+1. Since `u: U`, `u.type <: U`. (This is true, and useful, in the
+   scope of `u`, which is now invisible to the caller.)
+2. `iter` returns a `u.type#Needle`.
+    - Note: since `u` is not in scope for the caller, if we returned
+      this as is, it would effectively widen to the existentially
+      bound `u.type#Needle forSome {val u: U}`. But the same logic in
+      the next step would apply to that type.
+3. By the `#` left side covariance, `u.type#Needle` widens to
+   `U#Needle`.
+
+Pretty simple, by the standards of what we’ve seen so far.
+
+## Contravariance is the root of all…
+
+But things break down when we try to call `iter(n)`. Keep in mind that
+`n: U#Needle` and the expected type is `u.Needle`. Specifically: since
+we don’t know in the implementation that `U` is a singleton type, we
+can’t use the “singleton type equivalence” rule on it! But suppose
+that we *could*; that is, **suppose that we could constrain `U` to be
+a singleton type**.
+
+1. The argument type is `U#Needle`.
+2. By singleton equivalence, since `u: U` and `u` is stable, so
+   `u.type = U`.
+3. By substituting the left-hand side of the `#`, we get
+   `u.type#Needle`.
+4. This shortens to `u.Needle`.
+
+By our inability to constrain `U` in this way, though, we are
+restricted to places where `U` occurs in covariant position when using
+cake-extracted APIs, though. We can invoke functions like `init`,
+because they only have the singleton index occurring in covariant
+position. **We are out of luck with any functions with a contravariant
+or invariant occurrence of the index, though.** This is, well, most of
+them.
+
+## Extracting more types
+
+We have the same problem with the function version.
+
+```scala
+def stepTwiceHaystack[U <: LittleUniverse](
+    h: U#Haystack, n: U#Needle): U#Needle =
+  h.iter(h.iter(n))
```

Comment:
Cool! (I have tried `Singleton` in other cases, but couldn't get anywhere, so glad to see this.) I'll test it out later, and if all seems well rewrite to declare that all four escape paths are open :+1: